// vscode/vscode/src/vs/base/browser/ui/findinput/findInputToggles.js
const {
  getDefaultHoverDelegate
} = require('../hover/hoverDelegateFactory.js');
const {
  Toggle
} = require('../toggle/toggle.js');
const {
  Codicon
} = require('../../../common/codicons.js');
const {
  localizeWithPath
} = require('../../../../nls.js');
const _moduleId = "vs/base/browser/ui/findinput/findInputToggles";
const NLS_CASE_SENSITIVE_TOGGLE_LABEL = localizeWithPath(_moduleId, 0, "Match Case");
const NLS_WHOLE_WORD_TOGGLE_LABEL = localizeWithPath(_moduleId, 1, "Match Whole Word");
const NLS_REGEX_TOGGLE_LABEL = localizeWithPath(_moduleId, 2, "Use Regular Expression");
class CaseSensitiveToggle extends Toggle {
  constructor(opts) {
    super({
      icon: Codicon.caseSensitive,
      title: NLS_CASE_SENSITIVE_TOGGLE_LABEL + opts.appendTitle,
      isChecked: opts.isChecked,
      hoverDelegate: opts.hoverDelegate ?? getDefaultHoverDelegate('element'),
      inputActiveOptionBorder: opts.inputActiveOptionBorder,
      inputActiveOptionForeground: opts.inputActiveOptionForeground,
      inputActiveOptionBackground: opts.inputActiveOptionBackground
    });
  }
}
class WholeWordsToggle extends Toggle {
  constructor(opts) {
    super({
      icon: Codicon.wholeWord,
      title: NLS_WHOLE_WORD_TOGGLE_LABEL + opts.appendTitle,
      isChecked: opts.isChecked,
      hoverDelegate: opts.hoverDelegate ?? getDefaultHoverDelegate('element'),
      inputActiveOptionBorder: opts.inputActiveOptionBorder,
      inputActiveOptionForeground: opts.inputActiveOptionForeground,
      inputActiveOptionBackground: opts.inputActiveOptionBackground
    });
  }
}
class RegexToggle extends Toggle {
  constructor(opts) {
    super({
      icon: Codicon.regex,
      title: NLS_REGEX_TOGGLE_LABEL + opts.appendTitle,
      isChecked: opts.isChecked,
      hoverDelegate: opts.hoverDelegate ?? getDefaultHoverDelegate('element'),
      inputActiveOptionBorder: opts.inputActiveOptionBorder,
      inputActiveOptionForeground: opts.inputActiveOptionForeground,
      inputActiveOptionBackground: opts.inputActiveOptionBackground
    });
  }
}
exports.CaseSensitiveToggle = CaseSensitiveToggle;
exports.RegexToggle = RegexToggle;
exports.WholeWordsToggle = WholeWordsToggle;
