// vscode/vscode/src/vs/platform/theme/common/colors/menuColors.js
const {
  localizeWithPath
} = require('../../../../nls.js');
const {
  registerColor
} = require('../colorUtils.js');
const {
  contrastBorder,
  activeContrastBorder
} = require('./baseColors.js');
const {
  selectForeground,
  selectBackground
} = require('./inputColors.js');
const {
  listActiveSelectionForeground,
  listActiveSelectionBackground
} = require('./listColors.js');
const _moduleId = "vs/platform/theme/common/colors/menuColors";
const menuBorder = registerColor('menu.border', {
  dark: null,
  light: null,
  hcDark: contrastBorder,
  hcLight: contrastBorder
}, localizeWithPath(_moduleId, 0, "Border color of menus."));
const menuForeground = registerColor('menu.foreground', {
  dark: selectForeground,
  light: selectForeground,
  hcDark: selectForeground,
  hcLight: selectForeground
}, localizeWithPath(_moduleId, 1, "Foreground color of menu items."));
const menuBackground = registerColor('menu.background', {
  dark: selectBackground,
  light: selectBackground,
  hcDark: selectBackground,
  hcLight: selectBackground
}, localizeWithPath(_moduleId, 2, "Background color of menu items."));
const menuSelectionForeground = registerColor('menu.selectionForeground', {
  dark: listActiveSelectionForeground,
  light: listActiveSelectionForeground,
  hcDark: listActiveSelectionForeground,
  hcLight: listActiveSelectionForeground
}, localizeWithPath(_moduleId, 3, "Foreground color of the selected menu item in menus."));
const menuSelectionBackground = registerColor('menu.selectionBackground', {
  dark: listActiveSelectionBackground,
  light: listActiveSelectionBackground,
  hcDark: listActiveSelectionBackground,
  hcLight: listActiveSelectionBackground
}, localizeWithPath(_moduleId, 4, "Background color of the selected menu item in menus."));
const menuSelectionBorder = registerColor('menu.selectionBorder', {
  dark: null,
  light: null,
  hcDark: activeContrastBorder,
  hcLight: activeContrastBorder
}, localizeWithPath(_moduleId, 5, "Border color of the selected menu item in menus."));
const menuSeparatorBackground = registerColor('menu.separatorBackground', {
  dark: '#606060',
  light: '#D4D4D4',
  hcDark: contrastBorder,
  hcLight: contrastBorder
}, localizeWithPath(_moduleId, 6, "Color of a separator menu item in menus."));
exports.menuBackground = menuBackground;
exports.menuBorder = menuBorder;
exports.menuForeground = menuForeground;
exports.menuSelectionBackground = menuSelectionBackground;
exports.menuSelectionBorder = menuSelectionBorder;
exports.menuSelectionForeground = menuSelectionForeground;
exports.menuSeparatorBackground = menuSeparatorBackground;
