// vscode/vscode/src/vs/platform/userDataSync/common/userDataSync.service.js
const {
  createDecorator
} = require('../../instantiation/common/instantiation.js');
const IUserDataSyncStoreManagementService = createDecorator('IUserDataSyncStoreManagementService');
const IUserDataSyncStoreService = createDecorator('IUserDataSyncStoreService');
const IUserDataSyncLocalStoreService = createDecorator('IUserDataSyncLocalStoreService');
const IUserDataSyncEnablementService = createDecorator('IUserDataSyncEnablementService');
const IUserDataSyncService = createDecorator('IUserDataSyncService');
const IUserDataSyncResourceProviderService = createDecorator('IUserDataSyncResourceProviderService');
const IUserDataAutoSyncService = createDecorator('IUserDataAutoSyncService');
const IUserDataSyncUtilService = createDecorator('IUserDataSyncUtilService');
const IUserDataSyncLogService = createDecorator('IUserDataSyncLogService');
exports.IUserDataAutoSyncService = IUserDataAutoSyncService;
exports.IUserDataSyncEnablementService = IUserDataSyncEnablementService;
exports.IUserDataSyncLocalStoreService = IUserDataSyncLocalStoreService;
exports.IUserDataSyncLogService = IUserDataSyncLogService;
exports.IUserDataSyncResourceProviderService = IUserDataSyncResourceProviderService;
exports.IUserDataSyncService = IUserDataSyncService;
exports.IUserDataSyncStoreManagementService = IUserDataSyncStoreManagementService;
exports.IUserDataSyncStoreService = IUserDataSyncStoreService;
exports.IUserDataSyncUtilService = IUserDataSyncUtilService;
