// vscode/vscode/src/vs/editor/common/services/editorBaseApi.js
const {
  CancellationTokenSource
} = require('../../../base/common/cancellation.js');
const {
  Emitter
} = require('../../../base/common/event.js');
const {
  KeyChord
} = require('../../../base/common/keyCodes.js');
const {
  URI
} = require('../../../base/common/uri.js');
const {
  Position
} = require('../core/position.js');
const {
  Range
} = require('../core/range.js');
const {
  Selection
} = require('../core/selection.js');
const {
  Token
} = require('../languages.js');
const {
  KeyCode,
  SelectionDirection,
  MarkerSeverity,
  MarkerTag
} = require('../standalone/standaloneEnums.js');
class KeyMod {
  static {
    this.CtrlCmd = 2048;
  }
  static {
    this.Shift = 1024;
  }
  static {
    this.Alt = 512;
  }
  static {
    this.WinCtrl = 256;
  }
  static chord(firstPart, secondPart) {
    return KeyChord(firstPart, secondPart);
  }
}
function createMonacoBaseAPI() {
  return {
    editor: undefined,
    languages: undefined,
    CancellationTokenSource: CancellationTokenSource,
    Emitter: Emitter,
    KeyCode: KeyCode,
    KeyMod: KeyMod,
    Position: Position,
    Range: Range,
    Selection: Selection,
    SelectionDirection: SelectionDirection,
    MarkerSeverity: MarkerSeverity,
    MarkerTag: MarkerTag,
    Uri: URI,
    Token: Token
  };
}
exports.KeyMod = KeyMod;
exports.createMonacoBaseAPI = createMonacoBaseAPI;
