// vscode/vscode/src/vs/platform/theme/browser/defaultStyles.js
const {
  asCssVariable,
  asCssVariableWithDefault
} = require('../common/colorUtils.js');
const {
  contrastBorder,
  textLinkForeground,
  activeContrastBorder,
  focusBorder
} = require('../common/colors/baseColors.js');
const {} = require('../common/colors/chartsColors.js');
const {
  widgetShadow,
  editorWidgetBackground,
  editorWidgetForeground,
  problemsErrorIconForeground,
  problemsWarningIconForeground,
  problemsInfoIconForeground,
  breadcrumbsBackground,
  breadcrumbsForeground,
  breadcrumbsFocusForeground,
  breadcrumbsActiveSelectionForeground,
  editorWidgetBorder
} = require('../common/colors/editorColors.js');
const {
  keybindingLabelBackground,
  keybindingLabelForeground,
  keybindingLabelBorder,
  keybindingLabelBottomBorder,
  buttonForeground,
  buttonSeparator,
  buttonBackground,
  buttonHoverBackground,
  buttonSecondaryForeground,
  buttonSecondaryBackground,
  buttonSecondaryHoverBackground,
  buttonBorder,
  inputActiveOptionBorder,
  inputActiveOptionForeground,
  inputActiveOptionBackground,
  checkboxBackground,
  checkboxBorder,
  checkboxForeground,
  inputBackground,
  inputForeground,
  inputBorder,
  inputValidationInfoBorder,
  inputValidationInfoBackground,
  inputValidationInfoForeground,
  inputValidationWarningBorder,
  inputValidationWarningBackground,
  inputValidationWarningForeground,
  inputValidationErrorBorder,
  inputValidationErrorBackground,
  inputValidationErrorForeground,
  selectBackground,
  selectListBackground,
  selectForeground,
  selectBorder
} = require('../common/colors/inputColors.js');
const {
  listFilterWidgetBackground,
  listFilterWidgetOutline,
  listFilterWidgetNoMatchesOutline,
  listFilterWidgetShadow,
  listFocusBackground,
  listFocusForeground,
  listFocusOutline,
  listActiveSelectionBackground,
  listActiveSelectionForeground,
  listActiveSelectionIconForeground,
  listFocusAndSelectionOutline,
  listInactiveSelectionBackground,
  listInactiveSelectionIconForeground,
  listInactiveSelectionForeground,
  listInactiveFocusBackground,
  listInactiveFocusOutline,
  listHoverBackground,
  listHoverForeground,
  listDropOverBackground,
  listDropBetweenBackground,
  treeIndentGuidesStroke,
  treeInactiveIndentGuidesStroke,
  tableColumnsBorder,
  tableOddRowsBackgroundColor
} = require('../common/colors/listColors.js');
const {
  menuBorder,
  menuForeground,
  menuBackground,
  menuSelectionForeground,
  menuSelectionBackground,
  menuSelectionBorder,
  menuSeparatorBackground
} = require('../common/colors/menuColors.js');
const {} = require('../common/colors/minimapColors.js');
const {
  progressBarBackground,
  badgeBackground,
  badgeForeground,
  scrollbarShadow,
  scrollbarSliderBackground,
  scrollbarSliderHoverBackground,
  scrollbarSliderActiveBackground
} = require('../common/colors/miscColors.js');
const {
  pickerGroupForeground,
  quickInputListFocusBackground,
  quickInputListFocusIconForeground,
  quickInputListFocusForeground
} = require('../common/colors/quickpickColors.js');
const {} = require('../common/colors/searchColors.js');
const {
  Color
} = require('../../../base/common/color.js');
function overrideStyles(override, styles) {
  const result = {
    ...styles
  };
  for (const key in override) {
    const val = override[key];
    result[key] = val !== undefined ? asCssVariable(val) : undefined;
  }
  return result;
}
const defaultKeybindingLabelStyles = {
  keybindingLabelBackground: asCssVariable(keybindingLabelBackground),
  keybindingLabelForeground: asCssVariable(keybindingLabelForeground),
  keybindingLabelBorder: asCssVariable(keybindingLabelBorder),
  keybindingLabelBottomBorder: asCssVariable(keybindingLabelBottomBorder),
  keybindingLabelShadow: asCssVariable(widgetShadow)
};
const defaultButtonStyles = {
  buttonForeground: asCssVariable(buttonForeground),
  buttonSeparator: asCssVariable(buttonSeparator),
  buttonBackground: asCssVariable(buttonBackground),
  buttonHoverBackground: asCssVariable(buttonHoverBackground),
  buttonSecondaryForeground: asCssVariable(buttonSecondaryForeground),
  buttonSecondaryBackground: asCssVariable(buttonSecondaryBackground),
  buttonSecondaryHoverBackground: asCssVariable(buttonSecondaryHoverBackground),
  buttonBorder: asCssVariable(buttonBorder)
};
const defaultProgressBarStyles = {
  progressBarBackground: asCssVariable(progressBarBackground)
};
const defaultToggleStyles = {
  inputActiveOptionBorder: asCssVariable(inputActiveOptionBorder),
  inputActiveOptionForeground: asCssVariable(inputActiveOptionForeground),
  inputActiveOptionBackground: asCssVariable(inputActiveOptionBackground)
};
const defaultCheckboxStyles = {
  checkboxBackground: asCssVariable(checkboxBackground),
  checkboxBorder: asCssVariable(checkboxBorder),
  checkboxForeground: asCssVariable(checkboxForeground)
};
const defaultDialogStyles = {
  dialogBackground: asCssVariable(editorWidgetBackground),
  dialogForeground: asCssVariable(editorWidgetForeground),
  dialogShadow: asCssVariable(widgetShadow),
  dialogBorder: asCssVariable(contrastBorder),
  errorIconForeground: asCssVariable(problemsErrorIconForeground),
  warningIconForeground: asCssVariable(problemsWarningIconForeground),
  infoIconForeground: asCssVariable(problemsInfoIconForeground),
  textLinkForeground: asCssVariable(textLinkForeground)
};
const defaultInputBoxStyles = {
  inputBackground: asCssVariable(inputBackground),
  inputForeground: asCssVariable(inputForeground),
  inputBorder: asCssVariable(inputBorder),
  inputValidationInfoBorder: asCssVariable(inputValidationInfoBorder),
  inputValidationInfoBackground: asCssVariable(inputValidationInfoBackground),
  inputValidationInfoForeground: asCssVariable(inputValidationInfoForeground),
  inputValidationWarningBorder: asCssVariable(inputValidationWarningBorder),
  inputValidationWarningBackground: asCssVariable(inputValidationWarningBackground),
  inputValidationWarningForeground: asCssVariable(inputValidationWarningForeground),
  inputValidationErrorBorder: asCssVariable(inputValidationErrorBorder),
  inputValidationErrorBackground: asCssVariable(inputValidationErrorBackground),
  inputValidationErrorForeground: asCssVariable(inputValidationErrorForeground)
};
function getInputBoxStyle(override) {
  return overrideStyles(override, defaultInputBoxStyles);
}
const defaultFindWidgetStyles = {
  listFilterWidgetBackground: asCssVariable(listFilterWidgetBackground),
  listFilterWidgetOutline: asCssVariable(listFilterWidgetOutline),
  listFilterWidgetNoMatchesOutline: asCssVariable(listFilterWidgetNoMatchesOutline),
  listFilterWidgetShadow: asCssVariable(listFilterWidgetShadow),
  inputBoxStyles: defaultInputBoxStyles,
  toggleStyles: defaultToggleStyles
};
const defaultCountBadgeStyles = {
  badgeBackground: asCssVariable(badgeBackground),
  badgeForeground: asCssVariable(badgeForeground),
  badgeBorder: asCssVariable(contrastBorder)
};
const defaultBreadcrumbsWidgetStyles = {
  breadcrumbsBackground: asCssVariable(breadcrumbsBackground),
  breadcrumbsForeground: asCssVariable(breadcrumbsForeground),
  breadcrumbsHoverForeground: asCssVariable(breadcrumbsFocusForeground),
  breadcrumbsFocusForeground: asCssVariable(breadcrumbsFocusForeground),
  breadcrumbsFocusAndSelectionForeground: asCssVariable(breadcrumbsActiveSelectionForeground)
};
const defaultListStyles = {
  listBackground: undefined,
  listInactiveFocusForeground: undefined,
  listFocusBackground: asCssVariable(listFocusBackground),
  listFocusForeground: asCssVariable(listFocusForeground),
  listFocusOutline: asCssVariable(listFocusOutline),
  listActiveSelectionBackground: asCssVariable(listActiveSelectionBackground),
  listActiveSelectionForeground: asCssVariable(listActiveSelectionForeground),
  listActiveSelectionIconForeground: asCssVariable(listActiveSelectionIconForeground),
  listFocusAndSelectionOutline: asCssVariable(listFocusAndSelectionOutline),
  listFocusAndSelectionBackground: asCssVariable(listActiveSelectionBackground),
  listFocusAndSelectionForeground: asCssVariable(listActiveSelectionForeground),
  listInactiveSelectionBackground: asCssVariable(listInactiveSelectionBackground),
  listInactiveSelectionIconForeground: asCssVariable(listInactiveSelectionIconForeground),
  listInactiveSelectionForeground: asCssVariable(listInactiveSelectionForeground),
  listInactiveFocusBackground: asCssVariable(listInactiveFocusBackground),
  listInactiveFocusOutline: asCssVariable(listInactiveFocusOutline),
  listHoverBackground: asCssVariable(listHoverBackground),
  listHoverForeground: asCssVariable(listHoverForeground),
  listDropOverBackground: asCssVariable(listDropOverBackground),
  listDropBetweenBackground: asCssVariable(listDropBetweenBackground),
  listSelectionOutline: asCssVariable(activeContrastBorder),
  listHoverOutline: asCssVariable(activeContrastBorder),
  treeIndentGuidesStroke: asCssVariable(treeIndentGuidesStroke),
  treeInactiveIndentGuidesStroke: asCssVariable(treeInactiveIndentGuidesStroke),
  treeStickyScrollBackground: undefined,
  treeStickyScrollBorder: undefined,
  treeStickyScrollShadow: undefined,
  tableColumnsBorder: asCssVariable(tableColumnsBorder),
  tableOddRowsBackgroundColor: asCssVariable(tableOddRowsBackgroundColor)
};
function getListStyles(override) {
  return overrideStyles(override, defaultListStyles);
}
const defaultSelectBoxStyles = {
  selectBackground: asCssVariable(selectBackground),
  selectListBackground: asCssVariable(selectListBackground),
  selectForeground: asCssVariable(selectForeground),
  decoratorRightForeground: asCssVariable(pickerGroupForeground),
  selectBorder: asCssVariable(selectBorder),
  focusBorder: asCssVariable(focusBorder),
  listFocusBackground: asCssVariable(quickInputListFocusBackground),
  listInactiveSelectionIconForeground: asCssVariable(quickInputListFocusIconForeground),
  listFocusForeground: asCssVariable(quickInputListFocusForeground),
  listFocusOutline: asCssVariableWithDefault(activeContrastBorder, Color.transparent.toString()),
  listHoverBackground: asCssVariable(listHoverBackground),
  listHoverForeground: asCssVariable(listHoverForeground),
  listHoverOutline: asCssVariable(activeContrastBorder),
  selectListBorder: asCssVariable(editorWidgetBorder),
  listBackground: undefined,
  listActiveSelectionBackground: undefined,
  listActiveSelectionForeground: undefined,
  listActiveSelectionIconForeground: undefined,
  listFocusAndSelectionBackground: undefined,
  listDropOverBackground: undefined,
  listDropBetweenBackground: undefined,
  listInactiveSelectionBackground: undefined,
  listInactiveSelectionForeground: undefined,
  listInactiveFocusBackground: undefined,
  listInactiveFocusOutline: undefined,
  listSelectionOutline: undefined,
  listFocusAndSelectionForeground: undefined,
  listFocusAndSelectionOutline: undefined,
  listInactiveFocusForeground: undefined,
  tableColumnsBorder: undefined,
  tableOddRowsBackgroundColor: undefined,
  treeIndentGuidesStroke: undefined,
  treeInactiveIndentGuidesStroke: undefined,
  treeStickyScrollBackground: undefined,
  treeStickyScrollBorder: undefined,
  treeStickyScrollShadow: undefined
};
function getSelectBoxStyles(override) {
  return overrideStyles(override, defaultSelectBoxStyles);
}
const defaultMenuStyles = {
  shadowColor: asCssVariable(widgetShadow),
  borderColor: asCssVariable(menuBorder),
  foregroundColor: asCssVariable(menuForeground),
  backgroundColor: asCssVariable(menuBackground),
  selectionForegroundColor: asCssVariable(menuSelectionForeground),
  selectionBackgroundColor: asCssVariable(menuSelectionBackground),
  selectionBorderColor: asCssVariable(menuSelectionBorder),
  separatorColor: asCssVariable(menuSeparatorBackground),
  scrollbarShadow: asCssVariable(scrollbarShadow),
  scrollbarSliderBackground: asCssVariable(scrollbarSliderBackground),
  scrollbarSliderHoverBackground: asCssVariable(scrollbarSliderHoverBackground),
  scrollbarSliderActiveBackground: asCssVariable(scrollbarSliderActiveBackground)
};
exports.defaultBreadcrumbsWidgetStyles = defaultBreadcrumbsWidgetStyles;
exports.defaultButtonStyles = defaultButtonStyles;
exports.defaultCheckboxStyles = defaultCheckboxStyles;
exports.defaultCountBadgeStyles = defaultCountBadgeStyles;
exports.defaultDialogStyles = defaultDialogStyles;
exports.defaultFindWidgetStyles = defaultFindWidgetStyles;
exports.defaultInputBoxStyles = defaultInputBoxStyles;
exports.defaultKeybindingLabelStyles = defaultKeybindingLabelStyles;
exports.defaultListStyles = defaultListStyles;
exports.defaultMenuStyles = defaultMenuStyles;
exports.defaultProgressBarStyles = defaultProgressBarStyles;
exports.defaultSelectBoxStyles = defaultSelectBoxStyles;
exports.defaultToggleStyles = defaultToggleStyles;
exports.getInputBoxStyle = getInputBoxStyle;
exports.getListStyles = getListStyles;
exports.getSelectBoxStyles = getSelectBoxStyles;
