// @codingame/monaco-vscode-files-service-override/index.js
exports.CustomSchemas = CustomSchemas;
exports.DelegateFileSystemProvider = DelegateFileSystemProvider;
exports.EmptyFileSystemProvider = EmptyFileSystemProvider;
exports.OverlayFileSystemProvider = OverlayFileSystemProvider;
exports.RegisteredFile = RegisteredFile;
exports.RegisteredFileSystemProvider = RegisteredFileSystemProvider;
exports.RegisteredMemoryFile = RegisteredMemoryFile;
exports.RegisteredReadOnlyFile = RegisteredReadOnlyFile;
exports.createIndexedDBProviders = createIndexedDBProviders;
exports.default = "default";
exports.initFile = initFile;
exports.registerCustomProvider = registerCustomProvider;
exports.registerExtensionFile = registerExtensionFile;
exports.registerFileSystemOverlay = registerFileSystemOverlay;
exports.registerHTMLFileSystemProvider = registerHTMLFileSystemProvider;
exports.FileChangeType = FileChangeType;
exports.FilePermission = FilePermission;
exports.FileSystemProviderCapabilities = FileSystemProviderCapabilities;
exports.FileSystemProviderError = FileSystemProviderError;
exports.FileSystemProviderErrorCode = FileSystemProviderErrorCode;
exports.FileType = FileType;
exports.HTMLFileSystemProvider = HTMLFileSystemProvider;
exports.InMemoryFileSystemProvider = InMemoryFileSystemProvider;
exports.IndexedDB = IndexedDB;
exports.IndexedDBFileSystemProvider = IndexedDBFileSystemProvider;
