// vscode/vscode/src/vs/workbench/contrib/codeEditor/browser/toggleRenderControlCharacter.js
const {
  localize2WithPath,
  localizeWithPath
} = require('../../../../nls.js');
const {
  Action2,
  MenuId,
  registerAction2
} = require('../../../../platform/actions/common/actions.js');
const {
  IConfigurationService
} = require('../../../../platform/configuration/common/configuration.service.js');
const {
  ContextKeyExpr
} = require('../../../../platform/contextkey/common/contextkey.js');
const {
  Categories
} = require('../../../../platform/action/common/actionCommonCategories.js');
const _moduleId = "vs/workbench/contrib/codeEditor/browser/toggleRenderControlCharacter";
class ToggleRenderControlCharacterAction extends Action2 {
  static {
    this.ID = 'editor.action.toggleRenderControlCharacter';
  }
  constructor() {
    super({
      id: ToggleRenderControlCharacterAction.ID,
      title: {
        ...localize2WithPath(_moduleId, 0, "Toggle Control Characters"),
        mnemonicTitle: localizeWithPath(_moduleId, 1, "Render &&Control Characters")
      },
      category: Categories.View,
      f1: true,
      toggled: ContextKeyExpr.equals('config.editor.renderControlCharacters', true),
      menu: {
        id: MenuId.MenubarAppearanceMenu,
        group: '4_editor',
        order: 5
      }
    });
  }
  run(accessor) {
    const configurationService = accessor.get(IConfigurationService);
    const newRenderControlCharacters = !configurationService.getValue('editor.renderControlCharacters');
    return configurationService.updateValue('editor.renderControlCharacters', newRenderControlCharacters);
  }
}
registerAction2(ToggleRenderControlCharacterAction);
exports.ToggleRenderControlCharacterAction = ToggleRenderControlCharacterAction;
