// vscode/vscode/src/vs/editor/browser/widget/diffEditor/registrations.contribution.js
const {
  Codicon
} = require('../../../../base/common/codicons.js');
const {
  ThemeIcon
} = require('../../../../base/common/themables.js');
const {
  ModelDecorationOptions
} = require('../../../common/model/textModel.js');
const {
  localizeWithPath
} = require('../../../../nls.js');
const {
  registerColor
} = require('../../../../platform/theme/common/colorUtils.js');
const {} = require('../../../../platform/theme/common/colors/baseColors.js');
const {} = require('../../../../platform/theme/common/colors/chartsColors.js');
const {} = require('../../../../platform/theme/common/colors/editorColors.js');
const {} = require('../../../../platform/theme/common/colors/inputColors.js');
const {} = require('../../../../platform/theme/common/colors/listColors.js');
const {} = require('../../../../platform/theme/common/colors/menuColors.js');
const {} = require('../../../../platform/theme/common/colors/minimapColors.js');
const {} = require('../../../../platform/theme/common/colors/miscColors.js');
const {} = require('../../../../platform/theme/common/colors/quickpickColors.js');
const {} = require('../../../../platform/theme/common/colors/searchColors.js');
const {
  registerIcon
} = require('../../../../platform/theme/common/iconRegistry.js');
const _moduleId = "vs/editor/browser/widget/diffEditor/registrations.contribution";
registerColor('diffEditor.move.border', {
  dark: '#8b8b8b9c',
  light: '#8b8b8b9c',
  hcDark: '#8b8b8b9c',
  hcLight: '#8b8b8b9c'
}, localizeWithPath(_moduleId, 0, 'The border color for text that got moved in the diff editor.'));
registerColor('diffEditor.moveActive.border', {
  dark: '#FFA500',
  light: '#FFA500',
  hcDark: '#FFA500',
  hcLight: '#FFA500'
}, localizeWithPath(_moduleId, 1, 'The active border color for text that got moved in the diff editor.'));
registerColor('diffEditor.unchangedRegionShadow', {
  dark: '#000000',
  light: '#737373BF',
  hcDark: '#000000',
  hcLight: '#737373BF'
}, localizeWithPath(_moduleId, 2, 'The color of the shadow around unchanged region widgets.'));
const diffInsertIcon = registerIcon('diff-insert', Codicon.add, localizeWithPath(_moduleId, 3, 'Line decoration for inserts in the diff editor.'));
const diffRemoveIcon = registerIcon('diff-remove', Codicon.remove, localizeWithPath(_moduleId, 4, 'Line decoration for removals in the diff editor.'));
const diffLineAddDecorationBackgroundWithIndicator = ModelDecorationOptions.register({
  className: 'line-insert',
  description: 'line-insert',
  isWholeLine: true,
  linesDecorationsClassName: 'insert-sign ' + ThemeIcon.asClassName(diffInsertIcon),
  marginClassName: 'gutter-insert'
});
const diffLineDeleteDecorationBackgroundWithIndicator = ModelDecorationOptions.register({
  className: 'line-delete',
  description: 'line-delete',
  isWholeLine: true,
  linesDecorationsClassName: 'delete-sign ' + ThemeIcon.asClassName(diffRemoveIcon),
  marginClassName: 'gutter-delete'
});
const diffLineAddDecorationBackground = ModelDecorationOptions.register({
  className: 'line-insert',
  description: 'line-insert',
  isWholeLine: true,
  marginClassName: 'gutter-insert'
});
const diffLineDeleteDecorationBackground = ModelDecorationOptions.register({
  className: 'line-delete',
  description: 'line-delete',
  isWholeLine: true,
  marginClassName: 'gutter-delete'
});
const diffAddDecoration = ModelDecorationOptions.register({
  className: 'char-insert',
  description: 'char-insert',
  shouldFillLineOnLineBreak: true
});
const diffWholeLineAddDecoration = ModelDecorationOptions.register({
  className: 'char-insert',
  description: 'char-insert',
  isWholeLine: true
});
const diffAddDecorationEmpty = ModelDecorationOptions.register({
  className: 'char-insert diff-range-empty',
  description: 'char-insert diff-range-empty'
});
const diffDeleteDecoration = ModelDecorationOptions.register({
  className: 'char-delete',
  description: 'char-delete',
  shouldFillLineOnLineBreak: true
});
const diffWholeLineDeleteDecoration = ModelDecorationOptions.register({
  className: 'char-delete',
  description: 'char-delete',
  isWholeLine: true
});
const diffDeleteDecorationEmpty = ModelDecorationOptions.register({
  className: 'char-delete diff-range-empty',
  description: 'char-delete diff-range-empty'
});
exports.diffAddDecoration = diffAddDecoration;
exports.diffAddDecorationEmpty = diffAddDecorationEmpty;
exports.diffDeleteDecoration = diffDeleteDecoration;
exports.diffDeleteDecorationEmpty = diffDeleteDecorationEmpty;
exports.diffInsertIcon = diffInsertIcon;
exports.diffLineAddDecorationBackground = diffLineAddDecorationBackground;
exports.diffLineAddDecorationBackgroundWithIndicator = diffLineAddDecorationBackgroundWithIndicator;
exports.diffLineDeleteDecorationBackground = diffLineDeleteDecorationBackground;
exports.diffLineDeleteDecorationBackgroundWithIndicator = diffLineDeleteDecorationBackgroundWithIndicator;
exports.diffRemoveIcon = diffRemoveIcon;
exports.diffWholeLineAddDecoration = diffWholeLineAddDecoration;
exports.diffWholeLineDeleteDecoration = diffWholeLineDeleteDecoration;
