// vscode/vscode/src/vs/editor/contrib/gotoSymbol/browser/goToSymbol.js
const {
  coalesce
} = require('../../../../base/common/arrays.js');
const {
  CancellationToken
} = require('../../../../base/common/cancellation.js');
const {
  onUnexpectedExternalError
} = require('../../../../base/common/errors.js');
const {
  registerModelAndPositionCommand
} = require('../../../browser/editorExtensions.js');
const {
  ILanguageFeaturesService
} = require('../../../common/services/languageFeatures.js');
const {
  ReferencesModel
} = require('./referencesModel.js');
async function getLocationLinks(model, position, registry, provide) {
  const provider = registry.ordered(model);
  const promises = provider.map(provider => {
    return Promise.resolve(provide(provider, model, position)).then(undefined, err => {
      onUnexpectedExternalError(err);
      return undefined;
    });
  });
  const values = await Promise.all(promises);
  return coalesce(values.flat());
}
function getDefinitionsAtPosition(registry, model, position, token) {
  return getLocationLinks(model, position, registry, (provider, model, position) => {
    return provider.provideDefinition(model, position, token);
  });
}
function getDeclarationsAtPosition(registry, model, position, token) {
  return getLocationLinks(model, position, registry, (provider, model, position) => {
    return provider.provideDeclaration(model, position, token);
  });
}
function getImplementationsAtPosition(registry, model, position, token) {
  return getLocationLinks(model, position, registry, (provider, model, position) => {
    return provider.provideImplementation(model, position, token);
  });
}
function getTypeDefinitionsAtPosition(registry, model, position, token) {
  return getLocationLinks(model, position, registry, (provider, model, position) => {
    return provider.provideTypeDefinition(model, position, token);
  });
}
function getReferencesAtPosition(registry, model, position, compact, token) {
  return getLocationLinks(model, position, registry, async (provider, model, position) => {
    const result = await provider.provideReferences(model, position, {
      includeDeclaration: true
    }, token);
    if (!compact || !result || result.length !== 2) {
      return result;
    }
    const resultWithoutDeclaration = await provider.provideReferences(model, position, {
      includeDeclaration: false
    }, token);
    if (resultWithoutDeclaration && resultWithoutDeclaration.length === 1) {
      return resultWithoutDeclaration;
    }
    return result;
  });
}
async function _sortedAndDeduped(callback) {
  const rawLinks = await callback();
  const model = new ReferencesModel(rawLinks, '');
  const modelLinks = model.references.map(ref => ref.link);
  model.dispose();
  return modelLinks;
}
registerModelAndPositionCommand('_executeDefinitionProvider', (accessor, model, position) => {
  const languageFeaturesService = accessor.get(ILanguageFeaturesService);
  const promise = getDefinitionsAtPosition(languageFeaturesService.definitionProvider, model, position, CancellationToken.None);
  return _sortedAndDeduped(() => promise);
});
registerModelAndPositionCommand('_executeTypeDefinitionProvider', (accessor, model, position) => {
  const languageFeaturesService = accessor.get(ILanguageFeaturesService);
  const promise = getTypeDefinitionsAtPosition(languageFeaturesService.typeDefinitionProvider, model, position, CancellationToken.None);
  return _sortedAndDeduped(() => promise);
});
registerModelAndPositionCommand('_executeDeclarationProvider', (accessor, model, position) => {
  const languageFeaturesService = accessor.get(ILanguageFeaturesService);
  const promise = getDeclarationsAtPosition(languageFeaturesService.declarationProvider, model, position, CancellationToken.None);
  return _sortedAndDeduped(() => promise);
});
registerModelAndPositionCommand('_executeReferenceProvider', (accessor, model, position) => {
  const languageFeaturesService = accessor.get(ILanguageFeaturesService);
  const promise = getReferencesAtPosition(languageFeaturesService.referenceProvider, model, position, false, CancellationToken.None);
  return _sortedAndDeduped(() => promise);
});
registerModelAndPositionCommand('_executeImplementationProvider', (accessor, model, position) => {
  const languageFeaturesService = accessor.get(ILanguageFeaturesService);
  const promise = getImplementationsAtPosition(languageFeaturesService.implementationProvider, model, position, CancellationToken.None);
  return _sortedAndDeduped(() => promise);
});
exports.getDeclarationsAtPosition = getDeclarationsAtPosition;
exports.getDefinitionsAtPosition = getDefinitionsAtPosition;
exports.getImplementationsAtPosition = getImplementationsAtPosition;
exports.getReferencesAtPosition = getReferencesAtPosition;
exports.getTypeDefinitionsAtPosition = getTypeDefinitionsAtPosition;
