// vscode/vscode/src/vs/workbench/contrib/searchEditor/browser/constants.js
const {
  RawContextKey
} = require('../../../../platform/contextkey/common/contextkey.js');
const InSearchEditor = new RawContextKey('inSearchEditor', false);
const SearchEditorScheme = 'search-editor';
const SearchEditorWorkingCopyTypeId = 'search/editor';
const SearchEditorFindMatchClass = 'searchEditorFindMatch';
const SearchEditorID = 'workbench.editor.searchEditor';
const OpenNewEditorCommandId = 'search.action.openNewEditor';
const OpenEditorCommandId = 'search.action.openEditor';
const ToggleSearchEditorContextLinesCommandId = 'toggleSearchEditorContextLines';
const SearchEditorInputTypeId = 'workbench.editorinputs.searchEditorInput';
exports.InSearchEditor = InSearchEditor;
exports.OpenEditorCommandId = OpenEditorCommandId;
exports.OpenNewEditorCommandId = OpenNewEditorCommandId;
exports.SearchEditorFindMatchClass = SearchEditorFindMatchClass;
exports.SearchEditorID = SearchEditorID;
exports.SearchEditorInputTypeId = SearchEditorInputTypeId;
exports.SearchEditorScheme = SearchEditorScheme;
exports.SearchEditorWorkingCopyTypeId = SearchEditorWorkingCopyTypeId;
exports.ToggleSearchEditorContextLinesCommandId = ToggleSearchEditorContextLinesCommandId;
