// vscode/vscode/src/vs/workbench/services/userDataSync/common/userDataSync.js
const {
  RawContextKey,
  ContextKeyExpr
} = require('../../../../platform/contextkey/common/contextkey.js');
const {
  localize2WithPath,
  localizeWithPath
} = require('../../../../nls.js');
const {
  Codicon
} = require('../../../../base/common/codicons.js');
const {
  registerIcon
} = require('../../../../platform/theme/common/iconRegistry.js');
const {
  Categories
} = require('../../../../platform/action/common/actionCommonCategories.js');
const _moduleId = "vs/workbench/services/userDataSync/common/userDataSync";
function getSyncAreaLabel(source) {
  switch (source) {
    case "settings":
      return localizeWithPath(_moduleId, 0, "Settings");
    case "keybindings":
      return localizeWithPath(_moduleId, 1, "Keyboard Shortcuts");
    case "snippets":
      return localizeWithPath(_moduleId, 2, "User Snippets");
    case "tasks":
      return localizeWithPath(_moduleId, 3, "User Tasks");
    case "extensions":
      return localizeWithPath(_moduleId, 4, "Extensions");
    case "globalState":
      return localizeWithPath(_moduleId, 5, "UI State");
    case "profiles":
      return localizeWithPath(_moduleId, 6, "Profiles");
    case "workspaceState":
      return localizeWithPath(_moduleId, 7, "Workspace State");
  }
}
const SYNC_TITLE = localize2WithPath(_moduleId, 8, "Settings Sync");
const SYNC_VIEW_ICON = registerIcon('settings-sync-view-icon', Codicon.sync, localizeWithPath(_moduleId, 9, 'View icon of the Settings Sync view.'));
const CONTEXT_SYNC_STATE = new RawContextKey('syncStatus', "uninitialized");
const CONTEXT_SYNC_ENABLEMENT = new RawContextKey('syncEnabled', false);
const CONTEXT_ACCOUNT_STATE = new RawContextKey('userDataSyncAccountStatus', "unavailable");
const CONTEXT_ENABLE_ACTIVITY_VIEWS = new RawContextKey(`enableSyncActivityViews`, false);
const CONTEXT_ENABLE_SYNC_CONFLICTS_VIEW = new RawContextKey(`enableSyncConflictsView`, false);
const CONTEXT_HAS_CONFLICTS = new RawContextKey('hasConflicts', false);
const CONFIGURE_SYNC_COMMAND_ID = 'workbench.userDataSync.actions.configure';
const SHOW_SYNC_LOG_COMMAND_ID = 'workbench.userDataSync.actions.showLog';
const SYNC_VIEW_CONTAINER_ID = 'workbench.view.sync';
const SYNC_CONFLICTS_VIEW_ID = 'workbench.views.sync.conflicts';
const DOWNLOAD_ACTIVITY_ACTION_DESCRIPTOR = {
  id: 'workbench.userDataSync.actions.downloadSyncActivity',
  title: localize2WithPath(_moduleId, 10, "Download Settings Sync Activity"),
  category: Categories.Developer,
  f1: true,
  precondition: ContextKeyExpr.and(CONTEXT_ACCOUNT_STATE.isEqualTo("available"), CONTEXT_SYNC_STATE.notEqualsTo("uninitialized"))
};
exports.CONFIGURE_SYNC_COMMAND_ID = CONFIGURE_SYNC_COMMAND_ID;
exports.CONTEXT_ACCOUNT_STATE = CONTEXT_ACCOUNT_STATE;
exports.CONTEXT_ENABLE_ACTIVITY_VIEWS = CONTEXT_ENABLE_ACTIVITY_VIEWS;
exports.CONTEXT_ENABLE_SYNC_CONFLICTS_VIEW = CONTEXT_ENABLE_SYNC_CONFLICTS_VIEW;
exports.CONTEXT_HAS_CONFLICTS = CONTEXT_HAS_CONFLICTS;
exports.CONTEXT_SYNC_ENABLEMENT = CONTEXT_SYNC_ENABLEMENT;
exports.CONTEXT_SYNC_STATE = CONTEXT_SYNC_STATE;
exports.DOWNLOAD_ACTIVITY_ACTION_DESCRIPTOR = DOWNLOAD_ACTIVITY_ACTION_DESCRIPTOR;
exports.SHOW_SYNC_LOG_COMMAND_ID = SHOW_SYNC_LOG_COMMAND_ID;
exports.SYNC_CONFLICTS_VIEW_ID = SYNC_CONFLICTS_VIEW_ID;
exports.SYNC_TITLE = SYNC_TITLE;
exports.SYNC_VIEW_CONTAINER_ID = SYNC_VIEW_CONTAINER_ID;
exports.SYNC_VIEW_ICON = SYNC_VIEW_ICON;
exports.getSyncAreaLabel = getSyncAreaLabel;
