// vscode/vscode/src/vs/workbench/services/userDataProfile/common/userDataProfile.js
const {
  localizeWithPath,
  localize2WithPath
} = require('../../../../nls.js');
const {} = require('../../../../platform/actions/common/actions.js');
const {
  RawContextKey
} = require('../../../../platform/contextkey/common/contextkey.js');
const {
  URI
} = require('../../../../base/common/uri.js');
const {
  registerIcon
} = require('../../../../platform/theme/common/iconRegistry.js');
const {
  Codicon
} = require('../../../../base/common/codicons.js');
const _moduleId = "vs/workbench/services/userDataProfile/common/userDataProfile";
const PROFILE_URL_AUTHORITY = 'profile';
function toUserDataProfileUri(path, productService) {
  return URI.from({
    scheme: productService.urlProtocol,
    authority: PROFILE_URL_AUTHORITY,
    path: path.startsWith('/') ? path : `/${path}`
  });
}
const defaultUserDataProfileIcon = registerIcon('defaultProfile-icon', Codicon.settings, localizeWithPath(_moduleId, 0, 'Icon for Default Profile.'));
const PROFILES_TITLE = localize2WithPath(_moduleId, 1, 'Profiles');
const PROFILES_CATEGORY = {
  ...PROFILES_TITLE
};
const PROFILE_EXTENSION = 'code-profile';
const PROFILE_FILTER = [{
  name: localizeWithPath(_moduleId, 2, "Profile"),
  extensions: [PROFILE_EXTENSION]
}];
const CURRENT_PROFILE_CONTEXT = new RawContextKey('currentProfile', '');
const IS_PROFILE_EXPORT_IN_PROGRESS_CONTEXT = new RawContextKey('isProfileExportInProgress', false);
const IS_PROFILE_IMPORT_IN_PROGRESS_CONTEXT = new RawContextKey('isProfileImportInProgress', false);
exports.CURRENT_PROFILE_CONTEXT = CURRENT_PROFILE_CONTEXT;
exports.IS_PROFILE_EXPORT_IN_PROGRESS_CONTEXT = IS_PROFILE_EXPORT_IN_PROGRESS_CONTEXT;
exports.IS_PROFILE_IMPORT_IN_PROGRESS_CONTEXT = IS_PROFILE_IMPORT_IN_PROGRESS_CONTEXT;
exports.PROFILES_CATEGORY = PROFILES_CATEGORY;
exports.PROFILES_TITLE = PROFILES_TITLE;
exports.PROFILE_EXTENSION = PROFILE_EXTENSION;
exports.PROFILE_FILTER = PROFILE_FILTER;
exports.PROFILE_URL_AUTHORITY = PROFILE_URL_AUTHORITY;
exports.defaultUserDataProfileIcon = defaultUserDataProfileIcon;
exports.toUserDataProfileUri = toUserDataProfileUri;
