// vscode/vscode/src/vs/workbench/contrib/files/browser/fileConstants.js
const {
  localize2WithPath,
  localizeWithPath
} = require('../../../../nls.js');
const {
  RawContextKey
} = require('../../../../platform/contextkey/common/contextkey.js');
const _moduleId = "vs/workbench/contrib/files/browser/fileConstants";
const REVEAL_IN_EXPLORER_COMMAND_ID = 'revealInExplorer';
const REVERT_FILE_COMMAND_ID = 'workbench.action.files.revert';
const OPEN_TO_SIDE_COMMAND_ID = 'explorer.openToSide';
const OPEN_WITH_EXPLORER_COMMAND_ID = 'explorer.openWith';
const SELECT_FOR_COMPARE_COMMAND_ID = 'selectForCompare';
const COMPARE_SELECTED_COMMAND_ID = 'compareSelected';
const COMPARE_RESOURCE_COMMAND_ID = 'compareFiles';
const COMPARE_WITH_SAVED_COMMAND_ID = 'workbench.files.action.compareWithSaved';
const COPY_PATH_COMMAND_ID = 'copyFilePath';
const COPY_RELATIVE_PATH_COMMAND_ID = 'copyRelativeFilePath';
const SAVE_FILE_AS_COMMAND_ID = 'workbench.action.files.saveAs';
const SAVE_FILE_AS_LABEL = localize2WithPath(_moduleId, 0, "Save As...");
const SAVE_FILE_COMMAND_ID = 'workbench.action.files.save';
const SAVE_FILE_LABEL = localize2WithPath(_moduleId, 1, "Save");
const SAVE_FILE_WITHOUT_FORMATTING_COMMAND_ID = 'workbench.action.files.saveWithoutFormatting';
const SAVE_FILE_WITHOUT_FORMATTING_LABEL = localize2WithPath(_moduleId, 2, "Save without Formatting");
const SAVE_ALL_COMMAND_ID = 'saveAll';
const SAVE_ALL_LABEL = localize2WithPath(_moduleId, 3, "Save All");
const SAVE_ALL_IN_GROUP_COMMAND_ID = 'workbench.files.action.saveAllInGroup';
const SAVE_FILES_COMMAND_ID = 'workbench.action.files.saveFiles';
const OpenEditorsGroupContext = new RawContextKey('groupFocusedInOpenEditors', false);
const OpenEditorsDirtyEditorContext = new RawContextKey('dirtyEditorFocusedInOpenEditors', false);
const OpenEditorsReadonlyEditorContext = new RawContextKey('readonlyEditorFocusedInOpenEditors', false);
const ResourceSelectedForCompareContext = new RawContextKey('resourceSelectedForCompare', false);
const REMOVE_ROOT_FOLDER_COMMAND_ID = 'removeRootFolder';
const REMOVE_ROOT_FOLDER_LABEL = localizeWithPath(_moduleId, 4, "Remove Folder from Workspace");
const PREVIOUS_COMPRESSED_FOLDER = 'previousCompressedFolder';
const NEXT_COMPRESSED_FOLDER = 'nextCompressedFolder';
const FIRST_COMPRESSED_FOLDER = 'firstCompressedFolder';
const LAST_COMPRESSED_FOLDER = 'lastCompressedFolder';
const NEW_UNTITLED_FILE_COMMAND_ID = 'workbench.action.files.newUntitledFile';
const NEW_UNTITLED_FILE_LABEL = localize2WithPath(_moduleId, 5, "New Untitled Text File");
const NEW_FILE_COMMAND_ID = 'workbench.action.files.newFile';
exports.COMPARE_RESOURCE_COMMAND_ID = COMPARE_RESOURCE_COMMAND_ID;
exports.COMPARE_SELECTED_COMMAND_ID = COMPARE_SELECTED_COMMAND_ID;
exports.COMPARE_WITH_SAVED_COMMAND_ID = COMPARE_WITH_SAVED_COMMAND_ID;
exports.COPY_PATH_COMMAND_ID = COPY_PATH_COMMAND_ID;
exports.COPY_RELATIVE_PATH_COMMAND_ID = COPY_RELATIVE_PATH_COMMAND_ID;
exports.FIRST_COMPRESSED_FOLDER = FIRST_COMPRESSED_FOLDER;
exports.LAST_COMPRESSED_FOLDER = LAST_COMPRESSED_FOLDER;
exports.NEW_FILE_COMMAND_ID = NEW_FILE_COMMAND_ID;
exports.NEW_UNTITLED_FILE_COMMAND_ID = NEW_UNTITLED_FILE_COMMAND_ID;
exports.NEW_UNTITLED_FILE_LABEL = NEW_UNTITLED_FILE_LABEL;
exports.NEXT_COMPRESSED_FOLDER = NEXT_COMPRESSED_FOLDER;
exports.OPEN_TO_SIDE_COMMAND_ID = OPEN_TO_SIDE_COMMAND_ID;
exports.OPEN_WITH_EXPLORER_COMMAND_ID = OPEN_WITH_EXPLORER_COMMAND_ID;
exports.OpenEditorsDirtyEditorContext = OpenEditorsDirtyEditorContext;
exports.OpenEditorsGroupContext = OpenEditorsGroupContext;
exports.OpenEditorsReadonlyEditorContext = OpenEditorsReadonlyEditorContext;
exports.PREVIOUS_COMPRESSED_FOLDER = PREVIOUS_COMPRESSED_FOLDER;
exports.REMOVE_ROOT_FOLDER_COMMAND_ID = REMOVE_ROOT_FOLDER_COMMAND_ID;
exports.REMOVE_ROOT_FOLDER_LABEL = REMOVE_ROOT_FOLDER_LABEL;
exports.REVEAL_IN_EXPLORER_COMMAND_ID = REVEAL_IN_EXPLORER_COMMAND_ID;
exports.REVERT_FILE_COMMAND_ID = REVERT_FILE_COMMAND_ID;
exports.ResourceSelectedForCompareContext = ResourceSelectedForCompareContext;
exports.SAVE_ALL_COMMAND_ID = SAVE_ALL_COMMAND_ID;
exports.SAVE_ALL_IN_GROUP_COMMAND_ID = SAVE_ALL_IN_GROUP_COMMAND_ID;
exports.SAVE_ALL_LABEL = SAVE_ALL_LABEL;
exports.SAVE_FILES_COMMAND_ID = SAVE_FILES_COMMAND_ID;
exports.SAVE_FILE_AS_COMMAND_ID = SAVE_FILE_AS_COMMAND_ID;
exports.SAVE_FILE_AS_LABEL = SAVE_FILE_AS_LABEL;
exports.SAVE_FILE_COMMAND_ID = SAVE_FILE_COMMAND_ID;
exports.SAVE_FILE_LABEL = SAVE_FILE_LABEL;
exports.SAVE_FILE_WITHOUT_FORMATTING_COMMAND_ID = SAVE_FILE_WITHOUT_FORMATTING_COMMAND_ID;
exports.SAVE_FILE_WITHOUT_FORMATTING_LABEL = SAVE_FILE_WITHOUT_FORMATTING_LABEL;
exports.SELECT_FOR_COMPARE_COMMAND_ID = SELECT_FOR_COMPARE_COMMAND_ID;
